gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/avro/AvroKeyMapper.java

```diff
@@ -45,22 +45,60 @@
   protected void setup(Context context) throws IOException, InterruptedException {
     keySchema = AvroJob.getMapOutputKeySchema(context.getConfiguration());
     outKey = new AvroKey<GenericRecord>();
+    outKey.datum(new GenericData.Record(keySchema));
     outValue = new AvroValue<GenericRecord>();
   }
 
   @Override
-  protected void map(AvroKey<GenericRecord> key, NullWritable value, Context context) throws IOException,
-      InterruptedException {
+  protected void map(AvroKey<GenericRecord> key, NullWritable value, Context context)
+      throws IOException, InterruptedException {
     if (context.getNumReduceTasks() == 0) {
       context.write(key, NullWritable.get());
     } else {
-
-      // If there are reducers, mapper output key should be converted to the schema used for deduping, and
-      // mapper output value should be the original record.
-      outKey.datum(AvroUtils.convertRecordSchema(key.datum(), keySchema));
+      populateRecord(key.datum(), outKey.datum());
       outValue.datum(key.datum());
       context.write(outKey, outValue);
     }
   }
 
+  /**
+   * Populate the target record, based on the field values in the source record.
+   * Target record's schema should be a subset of source record's schema.
+   * Target record's schema cannot have MAP or ARRAY fields.
+   */
+  private static void populateRecord(GenericRecord source, GenericRecord target) {
```

Comment:
Can you rename this to `populateComparableKeyRecord` or something similar? The current name does not suggest that target Schema needs to be comparable, and it might be confusing in the future. (I know it's in the comments, but renaming the method is very easy).